src/borg/_hashindex.c

```diff
@@ -376,29 +379,69 @@ hashindex_get(HashIndex *index, const void *key)
     return BUCKET_ADDR(index, idx) + index->key_size;
 }
 
+inline int
+distance(HashIndex *index, int current_idx, int ideal_idx)
+{
+    /* If the current index is smaller than the ideal index we've wrapped
+       around the end of the bucket array and need to compensate for that. */
+    return current_idx - ideal_idx + ( (current_idx < ideal_idx) ? index->num_buckets : 0 );
+}
+
+inline void
+memswap(void *a, void *b, void *tmp, size_t entry_size) {
+    memcpy(tmp, a, entry_size);
+    memcpy(a, b, entry_size);
+    memcpy(b, tmp, entry_size);
+}
+
 static int
 hashindex_set(HashIndex *index, const void *key, const void *value)
 {
     int idx = hashindex_lookup(index, key);
-    uint8_t *ptr;
+    uint8_t *bucket_ptr;
+    int offset = 0;
+    int other_offset;
+    int entry_size = (index->key_size + index->value_size);
+    static void *entry_to_insert = NULL;
+    static void *tmp_entry = NULL;
+    if (entry_to_insert == NULL) {
+        entry_to_insert = malloc(entry_size * 2);
+        tmp_entry = entry_to_insert + entry_size;
+    }
     if(idx < 0)
     {
+        /* we don't have the key in the index we need to find an appropriate address */
         if(index->num_entries > index->upper_limit) {
+            /* we need to grow the hashindex */
             if(!hashindex_resize(index, grow_size(index->num_buckets))) {
                 return 0;
             }
         }
         idx = hashindex_index(index, key);
+        memcpy(entry_to_insert, key, index->key_size);
+        memcpy(entry_to_insert + index->key_size, value, index->value_size);
+        bucket_ptr = BUCKET_ADDR(index, idx);
         while(!BUCKET_IS_EMPTY(index, idx) && !BUCKET_IS_DELETED(index, idx)) {
+            /* we have a collision */
+            other_offset = distance(
+                index, idx, hashindex_index(index, bucket_ptr));
+            if(other_offset < offset) {
+                /* Swap the bucket at idx with the current entry_to_insert.
+                   This is the gist of robin-hood hashing, we rob from the key with the
+                   lower distance to its optimal address by swapping places with it. */
+                memswap(bucket_ptr, entry_to_insert, tmp_entry, entry_size);
+                offset = other_offset;
+            }
+            offset++;
             idx = (idx + 1) % index->num_buckets;
+            bucket_ptr = BUCKET_ADDR(index, idx);
         }
-        ptr = BUCKET_ADDR(index, idx);
-        memcpy(ptr, key, index->key_size);
-        memcpy(ptr + index->key_size, value, index->value_size);
+        memcpy(bucket_ptr, entry_to_insert, entry_size);
         index->num_entries += 1;
     }
     else
     {
+        /* we already have the key in the index we just need to update its value */
         memcpy(BUCKET_ADDR(index, idx) + index->key_size, value, index->value_size);
     }
     return 1;
```

Comment:
As I mentioned previously `hashindex_lookup` could take advantage of the
maximum offset for the hashindex, by giving up after looking at `maximum
offset` buckets. Since every `hashindex_set` does a lookup first this could
potentially be a great improvement for real world usage. However to track
maximum offset we will add some small overhead.

Deleting would be the tricky bit. If we are deleting a key with maximum
offset we don't know if there is another key with the same offset, or what
the next biggest offset is. But if we simply don't update maximum offset on
delete (basically maximum offset means maximum ever seen), we could still
reap a great benefit from it, and only re-compute it on hash resize.

The only requirement for doing this, as I said in a previous comment, would
be that this requires the hashindex be in robin_hood order already. One way
to achieve this is to re-insert every key when we read from disk. That way
we can keep things mutually compatible with old versions of borg. Or we can
introduce some sort of versioning scheme using the MAGIC string. Anyway, I
should have some time to implement it tonight or tomorrow and we can see if
it's worth further consideration.

On Sep 11, 2016 18:02, "TW" <notifications@github.com> wrote:

> In src/borg/_hashindex.c
> <https://github.com/borgbackup/borg/pull/1429#discussion_r78297662>:
>
> >          index->num_entries += 1;
> >      }
> >      else
> >      {
> > +        /* we already have the key in the index we just need to update its value */
> >          memcpy(BUCKET_ADDR(index, idx) + index->key_size, value, index->value_size);
> >      }
> >      return 1;
>
> i read through the rest of this function and it looks correct.
> have to check against the robin hood paper again, whether it is complete.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/borgbackup/borg/pull/1429/files/163c3a25f65b5d1077ee18480e3b3b504a5c6ab9#r78297662>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB0wZ69AwZnYP1szOjIY-H7fKOxWNbptks5qpCYagaJpZM4Ja5g->
> .
>